src/ui/public/kuery/ast/kql.peg

```diff
@@ -1,123 +1,150 @@
-/*
- * Kuery parser
- */
-
-/*
- * Initialization block
- */
+// Initialization block
 {
-  var nodeTypes = options.helpers.nodeTypes;
+  var { nodeTypes: { function: { buildNode }}} = options.helpers;
+}
+
+start
+  = OrQuery
+  / EmptyQuery
 
-  if (options.includeMetadata === undefined) {
-    options.includeMetadata = true;
+EmptyQuery
+  = Space* {
+    return buildNode('and', []);
   }
 
-  function addMeta(source, text, location) {
-    if (options.includeMetadata) {
-      return Object.assign(
-        {},
-        source,
-        {
-          text: text,
-          location: simpleLocation(location),
-        }
-      );
-    }
+OrQuery
+  = head:AndQuery Or tail:OrQuery {
+    return buildNode('or', [head, tail]);
+  }
+  / AndQuery
 
-    return source;
+AndQuery
+  = head:NotQuery And tail:AndQuery{
+    return buildNode('and', [head, tail]);
   }
+  / NotQuery
 
-  function simpleLocation(location) {
-    // Returns an object representing the position of the function within the expression,
-    // demarcated by the position of its first character and last character. We calculate these values
-    // using the offset because the expression could span multiple lines, and we don't want to deal
-    // with column and line values.
-    return {
-      min: location.start.offset,
-      max: location.end.offset
+NotQuery
+  = Not query:SubQuery {
+    return buildNode('not', query);
+  }
+  / SubQuery
+
+SubQuery
+  = '(' Space* query:OrQuery Space* ')' { return query; }
+  / Expression
+
+Expression
+  = FieldListExpression
+  / FieldRangeExpression
+  / ValueExpression
+
+FieldListExpression
+  = field:Literal Space* ':' Space* list:SubList {
+    function buildNodeWithField(field, list) {
```

Comment:
Is it necessary to build up this intermediate data structure and then loop through it like this? Couldn't we build the actual nodes in each of the `List` rules?

Reply:
We don't have access to the field itself in each of the `List` rules. I don't like this stuff either, though... It has some code smell. I'll play around with it and see what I can think of. Feel free to chime in if you have any ideas.

Reply:
f7c9a54